### ISSUE TITLE
memory leaks in index cursor

### ISSUE THREAD

**Issue by dmitri-shubin** [wiredtiger/wiredtiger#1474]
Hi!

Valgrind shows memory leaks in cursor opened on index.

Test:
```
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <wiredtiger.h>

#define WT_HOME "cursor_leak.db"

#define WT_CALL(call) \
    do { \
        const int __rc = call; \
        if (__rc != 0) { \
            fprintf(stderr, # call " at (%s:%d) failed: %s [%d]\n", __FILE__, __LINE__, wiredtiger_strerror(__rc), __rc); \
            exit(EXIT_FAILURE); \
        } \
    } while (0)

int
main(void)
{
        WT_CONNECTION *conn;
        WT_SESSION *session;
        WT_CURSOR *cur, *cur1;

        system("rm -rf " WT_HOME " && mkdir " WT_HOME);

        WT_CALL(wiredtiger_open(NULL, NULL, "create", &conn));
        WT_CALL(conn->open_session(conn, NULL, NULL, &session));

        WT_CALL(session->create(session, "table:access", "key_format=i,value_format=i,columns=(k,v)"));

        WT_CALL(session->create(session, "index:access:index", "columns=(v)"));

        WT_CALL(session->open_cursor(session, "index:access:index", NULL, NULL, &cur));
        WT_CALL(cur->close(cur));
        WT_CALL(session->close(session, NULL));
        WT_CALL(conn->close(conn, "leak_memory=no"));

        return 0;
}
```

Valgrind session:
```
$ LD_LIBRARY_PATH=../../.libs/ valgrind --tool=memcheck --leak-check=full --show-reachable=yes ./.libs/cursor_leak
==24714== Memcheck, a memory error detector
==24714== Copyright (C) 2002-2012, and GNU GPL'd, by Julian Seward et al.
==24714== Using Valgrind-3.8.1 and LibVEX; rerun with -h for copyright info
==24714== Command: ./.libs/cursor_leak
==24714==
==24714==
==24714== HEAP SUMMARY:
==24714==     in use at exit: 318 bytes in 9 blocks
==24714==   total heap usage: 1,434 allocs, 1,425 frees, 431,859 bytes allocated
==24714==
==24714== 2 bytes in 1 blocks are indirectly lost in loss record 1 of 9
==24714==    at 0x4C267BB: calloc (vg_replace_malloc.c:593)
==24714==    by 0x4EBB2FF: __wt_calloc (os_alloc.c:38)
==24714==    by 0x4EBB388: __wt_strndup (os_alloc.c:186)
==24714==    by 0x4EDC73E: __wt_schema_open_index (schema_open.c:225)
==24714==    by 0x4E90A01: __wt_curindex_open (cur_index.c:384)
==24714==    by 0x4EE98D4: __wt_open_cursor (session_api.c:259)
==24714==    by 0x4EE9C86: __session_open_cursor (session_api.c:351)
==24714==    by 0x400837: main (cursor_leak.c:35)
==24714==
==24714== 3 bytes in 1 blocks are indirectly lost in loss record 2 of 9
==24714==    at 0x4C267BB: calloc (vg_replace_malloc.c:593)
==24714==    by 0x4EBB2FF: __wt_calloc (os_alloc.c:38)
==24714==    by 0x4EBB388: __wt_strndup (os_alloc.c:186)
==24714==    by 0x4EDC4CB: __wt_schema_open_index (schema_open.c:155)
==24714==    by 0x4E90A01: __wt_curindex_open (cur_index.c:384)
==24714==    by 0x4EE98D4: __wt_open_cursor (session_api.c:259)
==24714==    by 0x4EE9C86: __session_open_cursor (session_api.c:351)
==24714==    by 0x400837: main (cursor_leak.c:35)
==24714==
==24714== 3 bytes in 1 blocks are indirectly lost in loss record 3 of 9
==24714==    at 0x4C267BB: calloc (vg_replace_malloc.c:593)
==24714==    by 0x4EBB2FF: __wt_calloc (os_alloc.c:38)
==24714==    by 0x4EBB388: __wt_strndup (os_alloc.c:186)
==24714==    by 0x4EDC780: __wt_schema_open_index (schema_open.c:234)
==24714==    by 0x4E90A01: __wt_curindex_open (cur_index.c:384)
==24714==    by 0x4EE98D4: __wt_open_cursor (session_api.c:259)
==24714==    by 0x4EE9C86: __session_open_cursor (session_api.c:351)
==24714==    by 0x400837: main (cursor_leak.c:35)
==24714== 
==24714== 4 bytes in 1 blocks are indirectly lost in loss record 4 of 9
==24714==    at 0x4C267BB: calloc (vg_replace_malloc.c:593)
==24714==    by 0x4EBB2FF: __wt_calloc (os_alloc.c:38)
==24714==    by 0x4EBB388: __wt_strndup (os_alloc.c:186)
==24714==    by 0x4EDC7E4: __wt_schema_open_index (schema_open.c:242)
==24714==    by 0x4E90A01: __wt_curindex_open (cur_index.c:384)
==24714==    by 0x4EE98D4: __wt_open_cursor (session_api.c:259)
==24714==    by 0x4EE9C86: __session_open_cursor (session_api.c:351)
==24714==    by 0x400837: main (cursor_leak.c:35)
==24714== 
==24714== 7 bytes in 1 blocks are indirectly lost in loss record 5 of 9
==24714==    at 0x4C267BB: calloc (vg_replace_malloc.c:593)
==24714==    by 0x4EBB2FF: __wt_calloc (os_alloc.c:38)
==24714==    by 0x4EBB388: __wt_strndup (os_alloc.c:186)
==24714==    by 0x4EDC6E4: __wt_schema_open_index (schema_open.c:219)
==24714==    by 0x4E90A01: __wt_curindex_open (cur_index.c:384)
==24714==    by 0x4EE98D4: __wt_open_cursor (session_api.c:259)
==24714==    by 0x4EE9C86: __session_open_cursor (session_api.c:351)
==24714==    by 0x400837: main (cursor_leak.c:35)
==24714== 
==24714== 19 bytes in 1 blocks are indirectly lost in loss record 6 of 9
==24714==    at 0x4C267BB: calloc (vg_replace_malloc.c:593)
==24714==    by 0x4EBB2FF: __wt_calloc (os_alloc.c:38)
==24714==    by 0x4EBB388: __wt_strndup (os_alloc.c:186)
==24714==    by 0x4EDC2A1: __wt_schema_open_index (schema_open.c:323)
==24714==    by 0x4E90A01: __wt_curindex_open (cur_index.c:384)
==24714==    by 0x4EE98D4: __wt_open_cursor (session_api.c:259)
==24714==    by 0x4EE9C86: __session_open_cursor (session_api.c:351)
==24714==    by 0x400837: main (cursor_leak.c:35)
==24714== 
==24714== 22 bytes in 1 blocks are indirectly lost in loss record 7 of 9
==24714==    at 0x4C267BB: calloc (vg_replace_malloc.c:593)
==24714==    by 0x4EBB2FF: __wt_calloc (os_alloc.c:38)
==24714==    by 0x4EBB388: __wt_strndup (os_alloc.c:186)
==24714==    by 0x4EDC3F6: __wt_schema_open_index (schema_open.c:143)
==24714==    by 0x4E90A01: __wt_curindex_open (cur_index.c:384)
==24714==    by 0x4EE98D4: __wt_open_cursor (session_api.c:259)
==24714==    by 0x4EE9C86: __session_open_cursor (session_api.c:351)
==24714==    by 0x400837: main (cursor_leak.c:35)
==24714==
==24714== 138 bytes in 1 blocks are indirectly lost in loss record 8 of 9
==24714==    at 0x4C267BB: calloc (vg_replace_malloc.c:593)
==24714==    by 0x4EBB2FF: __wt_calloc (os_alloc.c:38)
==24714==    by 0x4EBB388: __wt_strndup (os_alloc.c:186)
==24714==    by 0x4EDC2C5: __wt_schema_open_index (schema_open.c:324)
==24714==    by 0x4E90A01: __wt_curindex_open (cur_index.c:384)
==24714==    by 0x4EE98D4: __wt_open_cursor (session_api.c:259)
==24714==    by 0x4EE9C86: __session_open_cursor (session_api.c:351)
==24714==    by 0x400837: main (cursor_leak.c:35)
==24714==
==24714== 318 (120 direct, 198 indirect) bytes in 1 blocks are definitely lost in loss record 9 of 9
==24714==    at 0x4C267BB: calloc (vg_replace_malloc.c:593)
==24714==    by 0x4EBB2FF: __wt_calloc (os_alloc.c:38)
==24714==    by 0x4EDC281: __wt_schema_open_index (schema_open.c:322)
==24714==    by 0x4E90A01: __wt_curindex_open (cur_index.c:384)
==24714==    by 0x4EE98D4: __wt_open_cursor (session_api.c:259)
==24714==    by 0x4EE9C86: __session_open_cursor (session_api.c:351)
==24714==    by 0x400837: main (cursor_leak.c:35)
==24714==
==24714== LEAK SUMMARY:
==24714==    definitely lost: 120 bytes in 1 blocks
==24714==    indirectly lost: 198 bytes in 8 blocks
==24714==      possibly lost: 0 bytes in 0 blocks
==24714==    still reachable: 0 bytes in 0 blocks
==24714==         suppressed: 0 bytes in 0 blocks
==24714==
==24714== For counts of detected and suppressed errors, rerun with: -v
==24714== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 6 from 6)
```

Checked on 'develop' (d6f0c4cbab196fcf77b0f652fbca036ae770eeff)

**Comment by michaelcahill** [comment]
Hi @dmitri-shubin, this issue should be fixed on our develop branch by the merge of #1542.  That fix will be in our next release.

**Comment by dmitri-shubin** [comment]
Thanks @keithbostic, @michaelcahill !
